### ISSUE TITLE
raw dual iso export time!

### ISSUE THREAD

**Issue by gac123** [ilia3101/MLV-App#86]
mlv lite + sound raw dual iso 16bit 720p 25fps 01:38:33 coming from a 5d mkiii on a 1066x 128gb cf
intel xeon x5650 2.66ghz (6 cores/12threads), 16gb, evo 850 256gb, w10 x64, gtx1060

~37hours (dual iso on, interpolation AMaZE, alias on, blend on)
~26hours (dual iso on, interpolation AMaZE, alias off, blend off)

~15hours (dual iso on, interpolation mean, alias on, blend on)
~15hours (dual iso on, interpolation mean, alias on, blend oon) + processing (color correction)
~8hours (dual iso on, interpolation mean, alias off, blend off)

~3hours (dual iso off)

all with the same export settings (CinemaDNG uncompressed)

i'm shooting live concerts as a favor for a friend and was toying with magic lantern to maybe get an even better result, but with these processing times not even having it touched in premiere pro, it's a no-go

the app is not using any gpu, and barely my cpu (~8%)

i understand this process takes time but with many cores available it should be possible to assign multiple workers to make it more time-efficient by making better use of available resources

either way, great app :)

**Comment by masc4ii** [comment]
Thanks for the issue entry and your tests. 👍 
I did not know that this is so slow, because I never used this combination (only for some small tests).

CinemaDNG export runs on one thread only at the moment (@bouncyball-git : am I right, or is it already multithreaded? Could you please check that? Is the AMaZE algorithm the same as for debayering and multithreaded?).

The app uses no GPU - that is why it runs on nearly every computer instead of most other video apps. Maybe in far future we will support both - use GPU and don't use GPU as option. But until now we had no real success for using GPU. 😄 If you (or someone else) have skills in programming GPU that would be very welcome!

**Comment by masc4ii** [comment]
Again @bouncyball-git : is it possible to run multiple `saveDngFrame( m_pMlvObject, cinemaDng, frame, dngFileName.data() )` in parallel? I did that in v0.8 with the PNG export (before ffmpeg started conveting png to movie) using QThreadPool (see function `startExport()`). Could this be a way here? If I understand the llrawproc code right, it is not really made for multithreading...
If all other components on a 16 thread-CPU-computer are fast enough, that would nearly mean 37hours (see above) divided by 16 = 2:20hours... that would be enormous!

**Comment by masc4ii** [comment]
Would it be possible to get it thread safe with the other amaze code?
I am not sure if it is more thread safe when using the Qt solution... (I used a mutex, so I think it is not really better) 😢

**Comment by masc4ii** [comment]
No, never tried that... could also be a way.

**Comment by bouncyball-git** [comment]
Hey man I've been thinking about openmp too, but I had no experience with it so far.

Very nice that you tried it!!! Can I play with that your mod in linux?

**Comment by masc4ii** [comment]
Yes, you can try it out. For me linux worked so far...

**Comment by masc4ii** [comment]
On mac this could work... but I can't try it out here on my old 10.9 and XCode 6.1 ... works starting with XCode 6.2 brew tells (who knows if that's true).
https://stackoverflow.com/questions/44380459/is-openmp-available-in-high-sierra-llvm/47230419#47230419

**Comment by masc4ii** [comment]
@bouncyball-git : does it work for you on Linux? How much is the performance improved on your system? Is the any difference? On OSX I gave up for now. openmp comes with a special stdlib.h which is incompabile with the one we use for all variable definitions - so nothing else than the `#pragma ...` is working anymore when compiling 😆  😂

**Comment by bouncyball-git** [comment]
I'll try this evening. Had to do lot's of family stuff :). Even bought new refrigerator (BTW it's a German brand ;)).

**Comment by bouncyball-git** [comment]
@masc4ii 
Hey man I've tested OpenMP version on linux. I can admit it is ~2x faster. But...

Watch [this](http://bouncyball.myweb.ge/mlvapp/?dir=&download=OpenMP_Dual_Iso_quirk.mp4) video.

**Comment by masc4ii** [comment]
Autsch... looks not good. That happens if you stop playback?

**Comment by bouncyball-git** [comment]
Should we leave OpenMP is the code? I think it give about 30% speedup but works incorrectly as dicussed above. Maybe get rid of it?

**Comment by masc4ii** [comment]
Yapp. Reverted. https://github.com/ilia3101/MLV-App/commit/de1808f55fc6c368ef08ca3729932e25c6ba53c9

**Comment by bouncyball-git** [comment]
I seems now dualiso works w/o issues. pixelproc.c with openmp too.
Would be nice to try openmp pragma stuff on stripes and pattern noise and chroma smooth.

**Comment by masc4ii** [comment]
Often it is not possible to use the "pragma parallel for" for a for loop. This is the case, if there are 2 parameters in the for header, or if the cycles of the loop depends each other. 

Not possible:
```
for(int i=0, int y=0; i < 5; i++)
{
...
}
```

And also this does not work, because code can't run parallel.
```
int i = 0;
for( int y = 0; y < 5; y++)
{
    i++;
    a[y] = i;
}
```

**Comment by bouncyball-git** [comment]
There are none of this case as I remember.

**Comment by masc4ii** [comment]
In pixelproc.c I tried the pragma stuff. There are some "for" left, but they have this "add_pixel_to_map" - here it is not possible I think, because it could happen at the same time twice (a kind of the second case I wrote above).

Stripes.c:
it could work in stripes_apply_correction().

Patternnoise.c:
there are a lot of points where it could work.

Chroma_smooth.c could also work... 
Let's try out!

**Comment by masc4ii** [comment]
Try out https://github.com/ilia3101/MLV-App/commit/a031747d5e36a2c80946d6db3fb2d7e2a7988fb3
😊

**Comment by bouncyball-git** [comment]
As I tested all was working so far. Pattern noise sucks as always (slower then even hard dual iso processing)
Like the speed of forced bad pixel removal. Stripes also seem to work ok.

Thank you :)

**Comment by masc4ii** [comment]
What exactly does pattern noise? It looks always worse than without... I really don't know why using it. 
Yes, CS is faster... what a shame that this does not work on Mac.

**Comment by masc4ii** [comment]
If you like: take the dualiso with the pragma's and try out each single one on the reverted file. Maybe you'll find the one doing this stupid error... (but there are many pragma lines).

**Comment by bouncyball-git** [comment]
I've been tracking down the bug which I described in [this](http://bouncyball.myweb.ge/mlvapp/?dir=&download=OpenMP_Dual_Iso_quirk.mp4) video and this brought me to the frame caching part as always :wink:. Investigating further I could not determine the exact issue yet... that multythreading rock and sucks at the same time hahaha :smiley:

**Comment by Jip-Hop** [comment]
What I'm doing now is opening six instances of MLV App, all loading the same session XML file. Then from the first instance I export the first 1/6th of the clips, and on the second the second 1/6th of the clips etc. Takes a bit more time to setup but saves a lot of waiting time :) Works great on Windows, because there it opens another instance by default. On MacOS it will focus on the already opened instance of MLV App.

**Comment by masc4ii** [comment]
Hmmm... the "poor mans multithreading". 😄 The problem with dual iso is, that the current algorithm is 100% single threaded. We had no luck at all to change that easily.

**Comment by Jip-Hop** [comment]
I tried with MLVFS today, which converts to dng very quickly. But the Dual ISO results from MLV App are much better, so I'll stick with MLV App for Dual ISO. Faster Dual ISO export from MLV App would be awesome, even if it's just a version of parallel export: "poor mans multithreading" :p On the other hand, the combination of MLVFS for non Dual ISO and MLV App for the Dual ISO clips is not too time consuming either (assuming not too many Dual ISO shots).

**Comment by Jip-Hop** [comment]
Actually I think MLV App is so awesome, I'd love to use it as the only app to process MLV files. So I made a little Automator workflow (MacOS only) to help me with my "poor mans multithreading" 😄. It's used by selecting MLV files in the Finder, right clicking and then under Services choose "MLV App instances". It will evenly spread all the MLV files over as many MLV App instances as the computer has logical cores. For my MacBook, quad core with hyper threading, that means 8 threads and thus 8 MLV App instances. Then I have to manually start the export 8 times. Took exporting 22 MLV files down from 23 minutes to 8 minutes 🙂 As the next step I'd like to look into a post processing script to automatically apply white balance corrections to the exported DNG files with dcraw, like what @dannephoto does with his Switch app.

<img width="630" alt="Screenshot 2020-08-16 at 01 10 33" src="https://user-images.githubusercontent.com/2871973/90324082-a0787e00-df6a-11ea-9a11-2144e5e4427d.png">

[MLV App instances.workflow.zip](https://github.com/ilia3101/MLV-App/files/5079798/MLV.App.instances.workflow.zip)

**Comment by dannephoto** [comment]
Regarding spreading evenly. It would be cool if all could be done in command line bypassing opening files, applying all raw settings before export.

**Comment by Jip-Hop** [comment]
Well, the GUI needs to be opened at some point to actually tweak the raw settings (and possibly Dual ISO settings) for each file. Could be done once and saved to a session file. Then the clips in this session file could be spread evenly a cross multiple MLV App instances, and if the command line would support it start exporting automatically. But at this point I see no benefit to mess with the command line any more :p and it might as well have been a button in MLV App: "Spawn multiple instances and export".

**Comment by masc4ii** [comment]
Why not defining a user default marxml receipt, and then just open the files. All should automatically do the same.

**Comment by Jip-Hop** [comment]
I think that option would work with my workflow file :) currently I'm fine with the default settings mostly. Except manually enabling Dual ISO where appropriate.

**Comment by Jip-Hop** [comment]
Reading this thread I was under the impression that DNG output wasn't multithreaded. Not just for Dual ISO, but for DNG export in general. But it looks like it actually is multithreaded for non Dual ISO DNG export:

<img width="419" alt="Screenshot 2020-08-16 at 11 34 12" src="https://user-images.githubusercontent.com/2871973/90332331-34bf0100-dfbc-11ea-9152-b4434542ed73.png">

However, even with multithreading this single MLV App instance took 51 minutes in total to export 110 MLV files (non Dual ISO) to uncompressed DNG, no sound.

Testing the multi instance export with the same export settings it only took 18 minutes in total.

In both of my tests, on two different computers, the speed with multiple instances seems to be about 2.8x faster.

So even though it may seem silly to open multiple instances for export, it actually is much faster 😄

**Comment by masc4ii** [comment]
If the algorithms allow, MLVApp is multithreaded. E.g. DualISO doesn't allow. If the algorithms allow multithreading, you'll have CPU usage overhead for controlling the threads. That is not just threory 😄

**Comment by dannephoto** [comment]
The more logical way is to be able and set all mlv_dump settings from command line without opening 12 gui copies and the need to push export button for each one. At least having the starting of the apps automated wojld already be much nicer.

**Comment by Jip-Hop** [comment]
I got the same results with just 6 instances, instead of 12. So I think limiting the amount of instances to the amount of physical cores (instead of logical cores) would be good enough. Saves clicking the export button a couple of times :P

**Comment by Gabboxl** [comment]
Hi, recently I started using Dual ISO when recording ML videos with my camera. As many people have noticed, the image processing in MLVApp is very slow. First of all, the biggest problem is that the rendering phase does not use the full potential of the CPU.

Many times it has been said that the Dual ISO algorithm and the DNG export are not thread safe, thus not parallelizable. I took a look at the code, but it looks like they are actually thread safe (the DNG export needed just a little change).

I sent 2 pull requests (linked above) that speed up Dual ISO processing and DNG export by parallelizing the frames rendering, and I didn't notice any differences in the final output. (DNG export on my little PC is at least 2x faster)

I took a look at the https://github.com/anibarro/MLV-App fork (linked at https://github.com/ilia3101/MLV-App/issues/286#issuecomment-2460633645), and I must say that this is a very wrong approach to making things faster in general, especially when using OpenMP.

Basically, what has been done in that fork (and talked about in this issue) was making EVERY SINGLE for loop parallel using OpenMP directives. That just pollutes the code, creates unnecessary little race conditions (difficult to debug and to be made thread safe), and creates A LOT of overhead with implicit barriers and small loops parallelizations.

The correct approach is to consider the entire processing pipeline and make the parallel sections as big as they can be; this is where we can benefit from parallelization the most.


Looking forward to people trying out the changes and seeing if there are issues on other platforms, as I tested them only on Windows.
